library/core/src/num/nonzero.rs

```diff
@@ -287,6 +287,204 @@ nonzero_integers_div! {
+}
+
+macro_rules! nonzero_integers_mul {
+    ( $( $Ty: ident($Int: ty); )+ ) => {
+        $(
+            impl $Ty {
+                /// Multiply two non-zero integers together.
+                /// Return [`None`] on overflow or if `other` is zero.
+                ///
+                /// # Examples
+                ///
+                /// ```
+                /// # #![feature(nonzero_mul)]
+                #[doc = concat!("# use std::num::", stringify!($Ty), ";")]
+                #[doc = concat!("let two = ", stringify!($Ty), "::new(2).unwrap();")]
+                #[doc = concat!("let four = ", stringify!($Ty), "::new(4).unwrap();")]
+                #[doc = concat!("let max = ", stringify!($Ty), "::new(",
+                                stringify!($Int), "::MAX).unwrap();")]
+                /// assert_eq!(Some(four), two.checked_mul(two));
+                /// assert_eq!(None, max.checked_mul(two));
+                /// ```
+                #[unstable(feature = "nonzero_mul", issue = "none")]
+                #[inline]
+                pub fn checked_mul(self, other: $Ty) -> Option<$Ty> {
+                    self.get().checked_mul(other.get()).map(|result| {
+                        // SAFETY: checked_mul returns None on overflow
+                        // and `other` is also non-null
+                        // so the result cannot be zero.
+                        unsafe { $Ty::new_unchecked(result) }
+                    })
+                }
+
+                /// Multiply two non-zero integers together.
+                #[doc = concat!("Return [`", stringify!($Int), "::MAX`] on overflow.")]
+                ///
+                /// # Examples
+                ///
+                /// ```
+                /// # #![feature(nonzero_mul)]
+                #[doc = concat!("# use std::num::", stringify!($Ty), ";")]
+                #[doc = concat!("let two = ", stringify!($Ty), "::new(2).unwrap();")]
+                #[doc = concat!("let four = ", stringify!($Ty), "::new(4).unwrap();")]
+                #[doc = concat!("let max = ", stringify!($Ty), "::new(",
+                                stringify!($Int), "::MAX).unwrap();")]
+                /// assert_eq!(four, two.saturating_mul(two));
+                /// assert_eq!(max, four.saturating_mul(max));
+                /// ```
+                #[unstable(feature = "nonzero_mul", issue = "none")]
+                #[inline]
+                pub fn saturating_mul(self, other: $Ty) -> $Ty {
+                    // SAFETY: saturating_mul returns u*::MAX on overflow
+                    // and `other` is also non-null
+                    // so the result cannot be zero.
+                    unsafe { $Ty::new_unchecked(self.get().saturating_mul(other.get())) }
+                }
+            }
+        )+
+    }
+}
+
+nonzero_integers_mul! {
+    NonZeroU8(u8);
+    NonZeroU16(u16);
+    NonZeroU32(u32);
+    NonZeroU64(u64);
+    NonZeroU128(u128);
+    NonZeroUsize(usize);
+}
```

Comment:
How about signed types? (Same question for `pow`.)

Reply:
Very true, signed nonzero types cannot be zero when multiplied by other nonzero values or raised to integer powers unless they overflow. I'm adding them :)

Reply:
> (You can use the same feature gate for all of them, maybe `nonzero_checked_ops` or something like that.)

@m-ou-se Okay. Since you also offer to implement `mul` and `pow` for signed nonzeroes, I'll get a little more ambitious then and implement other "ops" like `neg`, maybe `next_power_of_two` and other things I find in the std doc for `usize`.

I have no particular inclusion criterion yet (any idea?), but I'll exclude every operation for which it is *possible to get a null result null result without overflowing the type*. This rules out `div`, `rem`, `sub`, *etc.*

I'll make sure all tests pass and the new "ops" undergo revision again before I open the tracking issue.